### PR TITLE
Enable History datatype for Sync

### PR DESCRIPTION
**Pull request by AlexeyBarabash**
Resolves https://github.com/brave/brave-browser/issues/32876

This PR enables History datatype in Brave, becasue after CR119 the old TypedUrls datatype, which was used for History Sync, is wiped out from Chromium codebase.

Related server PR https://github.com/brave/go-sync/pull/176 .

Key points of the PR:
- reverts the PRs which were using TypedUrl: PR#16705 (Sync all history),  PR#16478 (Display history sync diagnostic info), PR#20071 (Disable history sync datatype to fallback to typed urls datatype);
- Overrides `HistoryModelTypeController` and `HistoryDeleteDirectivesModelTypeController` to allow start when encrypt everything option is enabled;
- HISTORY datatype is included into encrypteble types, so only clients in the chain can see raw urls, server can't; 
- HISTORY_DELETE_DIRECTIVES is kept unencrypted, because it is a technical data; `HistoryDeleteDirectiveSpecifics::UrlDirective::url` field is forcibly cleaned out not to leak, **chrome.history.deleteUrl extension API will not work properly**;
- modified `LowPriorityUserTypes` - `HISTORY` is removed, `HISTORY_DELETE_DIRECTIVES` added. We need this to make device first download History and then delete directives download and apply after. This patch will be reverted when the server will support delete directives.

Note: history entries created before enabling sync and turn history type on will not be sent to the server - inherited from upstream behavior .

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested] . Requested https://github.com/brave/reviews/issues/1427 
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Please note, this PR requires updated server, either built from https://github.com/brave/go-sync/pull/176 or wait till it will be merged/deployed.

**I. Browse/manual remove history entries**
1. Enable Sync on deviceA, turn on History datatype;
2. Join deviceB to the chain, also turn on History datatype;
3. From deviceA visit urls
https://search.brave.com/search?q=1000001&source=web
https://search.brave.com/search?q=1000002&source=web
https://search.brave.com/search?q=1000003&source=web
4. From deviceB visit urls
https://search.brave.com/search?q=2000001&source=web
https://search.brave.com/search?q=2000002&source=web
https://search.brave.com/search?q=2000003&source=web
5. Ensure both devices have those urls in history. Maybe need to wait ~1 minute max.
6. From deviceA remove https://search.brave.com/search?q=2000002&source=web url from history; from deviceB remove https://search.brave.com/search?q=1000002&source=web url from history; 
7. Wait a little (< 1min), refresh history pages. Expected not to have https://search.brave.com/search?q=2000002&source=web and https://search.brave.com/search?q=1000002&source=web on both devices.
8. Join deviceC to the sync chain, enable history
9. Check brave://history page on deviceC - expected not to have urls https://search.brave.com/search?q=2000002&source=web , https://search.brave.com/search?q=1000002&source=web

**II. Time-based history cleanup**
You may use what you have left from test (I), or you join two devices to the sync chain, enable history, and do some browsing.

1. open Clear browsing data dialog, brave://settings/clearBrowserData .
2. On deviceA choose only `Browsing history` and `Time range` `Last hour`, press `Clear data`
3. Ensure corresponding history entries were deleted on other devices.

**III. Ensure url from `chrome.history.deleteUrl` is not leaked**. This is test plan for developers or for security engineers, imho QA engineers may not go over the steps below

1. Run browser with command line `./brave --enable-logging=stderr --v=0  --vmodule=commit.cc=3,*/sync/engine/*=3,traffic_logger=1` from terminal;
2. Enable/join sync chain, enable history;
3. Open page brave://inspect/#extensions
4. Click `inspect` under Brave extension
5. Paste at console
```
chrome.history.deleteUrl({url: 'https://developer.chrome.com/docs/extensions/reference/history/#method-search'}, function() {console.log("deleted");})
```
6. search for ` "url_directive": {` in terminal.
Expected not to find such records.

**IV. Android**
1. Join Android device to chain, enable history datatype;
2. Expected to sync History between Android device and other devices in chain.

### CODE REVIEW

**Comment by AlexeyBarabash** [comment]
Removed `CI/Skip` to get some installers for draft PR

File: components/history/core/browser/sync/brave_history_delete_directives_model_type_controller.cc [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+                                                 model_type_store_service,
+                                                 history_service,
+                                                 pref_service) {}
+
+BraveHistoryDeleteDirectivesModelTypeController::
+    ~BraveHistoryDeleteDirectivesModelTypeController() = default;
+
+syncer::DataTypeController::PreconditionState
+BraveHistoryDeleteDirectivesModelTypeController::GetPreconditionState() const {
+  DCHECK(CalledOnValidThread());
```
[the remark this answers is not in the document]
Comment (fmarier) [reply components/history/core/browser/sync/brave_history_delete_directives_model_type_controller.cc:30]: False positive.
https://github.com/brave/security-action/issues/400

File: patches/components-history-core-browser-sync-delete_directive_handler.cc.patch [hunk 2]
```diff
@@ -0,0 +1,12 @@
+diff --git a/components/history/core/browser/sync/delete_directive_handler.cc b/components/history/core/browser/sync/delete_directive_handler.cc
+index aed3aec1f04f15f9f806887110d87abac9ef06ba..13053b7fe8a0281e3f9d9f69b58bf35e5d199e2e 100644
+--- a/components/history/core/browser/sync/delete_directive_handler.cc
++++ b/components/history/core/browser/sync/delete_directive_handler.cc
+@@ -387,7 +387,6 @@ bool DeleteDirectiveHandler::CreateUrlDeleteDirective(const GURL& url) {
```
Comment (bridiver) [inline patches/components-history-core-browser-sync-delete_directive_handler.cc.patch:5]: chromium_src override the entire method and return false

File: patches/components-sync-base-model_type.h.patch [hunk 3]
```diff
@@ -0,0 +1,13 @@
+diff --git a/components/sync/base/model_type.h b/components/sync/base/model_type.h
+index 8e423e993701c104d4a669c04a46aaa9beb93cc9..2c44991316671a4b8b207e451a71183debb5c66d 100644
+--- a/components/sync/base/model_type.h
++++ b/components/sync/base/model_type.h
```
[the remark this answers is not in the document]
Comment (bridiver) [reply patches/components-sync-base-model_type.h.patch:4]: please add a comment and issue link for processing these on the server so this gets deleted when that happens

File: patches/components-history-core-browser-BUILD.gn.patch [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
-index fcd97b996f10aca76a8406fc4058ce69355ca87e..5e64c0d27c4f9473eed3b077685b907d17149075 100644
+index fcd97b996f10aca76a8406fc4058ce69355ca87e..30e37c1586ee129a57d7db17cf9a7f913d75b8cf 100644
 --- a/components/history/core/browser/BUILD.gn
 +++ b/components/history/core/browser/BUILD.gn
 @@ -157,6 +157,7 @@ static_library("browser") {
        "android/visit_sql_handler.h",
      ]
    }
-+  import("//brave/components/history/core/browser/sources.gni") sources += brave_components_history_core_browser_sources deps += brave_components_history_core_browser_deps
++  import("//brave/components/history/core/browser/sync/sources.gni") sources += brave_components_history_core_browser_sync_sources
```
[the remark this answers is not in the document]
Comment (AlexeyBarabash) [reply patches/components-history-core-browser-BUILD.gn.patch]: thanks for the detailed explanation,

fixed at https://github.com/brave/brave-core/pull/20329/commits/72cc7e92c0a132c509aefa886ad8d7622efd3e49

File: patches/components-history-core-browser-sync-delete_directive_handler.cc.patch (hunk 2 above)
Comment (AlexeyBarabash) [reply patches/components-history-core-browser-sync-delete_directive_handler.cc.patch]: fixed at https://github.com/brave/brave-core/pull/20329/commits/fa0fe5fc915c0d66b13ff2e16a1c7e27c11118b5

File: patches/components-sync-base-model_type.h.patch (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexeyBarabash) [reply patches/components-sync-base-model_type.h.patch]: fixed at https://github.com/brave/brave-core/pull/20329/commits/0db92abfc172d73a1f6e89e8e20166ec8a2aef32

issue link is https://github.com/brave/go-sync/issues/178

File: chromium_src/components/sync/base/model_type.cc [hunk 5]
```diff
@@ -14,6 +14,7 @@ ModelTypeSet EncryptableUserTypes() {
   ModelTypeSet encryptable_user_types = EncryptableUserTypes_ChromiumImpl();
   // Brave sync has encryption setup ready when sync chain created
   encryptable_user_types.Put(DEVICE_INFO);
+  encryptable_user_types.Put(HISTORY);
```
Comment (bridiver) [inline chromium_src/components/sync/base/model_type.cc:17]: don't we need to remove `HISTORY_DELETE_DIRECTIVES` from this list?

Comment (AlexeyBarabash) [reply chromium_src/components/sync/base/model_type.cc:17]: [EncryptableUserTypes](https://source.chromium.org/chromium/chromium/src/+/main:components/sync/base/model_type.cc;l=583?q=EncryptableUserTypes&ss=chromium%2Fchromium%2Fsrc)
already has 
```
encryptable_user_types.Remove(HISTORY_DELETE_DIRECTIVES);
```
so I didn't put it here.

**Review by bridiver — APPROVED** [review]

**Pull request merged**